### PR TITLE
Side effects for user defined functions

### PR DESCRIPTION
**Pull request by chriseth**
Extracted from https://github.com/ethereum/solidity/pull/7137

Still need to take a look at #7137 if CommonSubexpressionEliminator is really the only component that can benefit from the side-effects.

closes #7137

### CODE REVIEW

**Comment by chriseth** [comment]
Argh, wanted to make this a draft.

File: test/libyul/yulOptimizerTests/commonSubexpressionEliminator/movable_functions.yul [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+}
+// ====
+// step: commonSubexpressionEliminator
+// ----
+// {
+//     function double(x) -> y
+//     { y := add(x, x) }
+//     let a := double(7)
+//     let b := 7
+//     let c := double(b)
```
Comment (chriseth) [inline test/libyul/yulOptimizerTests/commonSubexpressionEliminator/movable_functions.yul]: This should be `let c := a`, but maybe I have to use a literal `7`.

**Comment by chriseth** [comment]
Needs more tests that show that side-effects are actually properly computed, especially with recursive calls etc.

File: test/cmdlineTests/standard_eWasm_requested/output.json [hunk 2]
```diff
@@ -147,16 +133,6 @@
     (get_local $y)
 )
 
-(func $endian_swap
```
Comment (chriseth) [inline test/cmdlineTests/standard_eWasm_requested/output.json:150]: Removal of a pure function - I would say this looks good even though I would need to see the yul code to actually find out what happened here.

**Comment by chriseth** [comment]
The load resolver should also make use of this.

**Comment by chriseth** [comment]
Actually no, the load resolver should not use this.

**Comment by chriseth** [comment]
RedundantAssignEliminator might make use of it, but if it is run after SSATransform, the right-hand-sides of assignments (with a single variable) are variables anyway, so it is not helpful.

The UnusedPruner should use side-effects of functions instead.

**Comment by chriseth** [comment]
The required tests mentioned above are already merged (3c4f55824)

**Comment by chriseth** [comment]
Re-correcting myself: The load resolver can make heavy use of this, because it should not clear knowledge about storage/memory when functions are called that do not touch memory/storage!

**Comment by ekpyron** [comment]
Why not make the constructor of ``CommonSubexpressionEliminator`` private and replace it by a static ``run`` on the AST that itself calls ``SideEffectsPropagator::sideEffects(_dialect, CallGraphGenerator::callGraph(ast))`` and passes the result down to the private constructor and from there down to ``DataFlowAnalyzer`` by value?

**Comment by chriseth** [comment]
Updated.

File: libyul/optimiser/CommonSubexpressionEliminator.cpp [hunk 3]
```diff
@@ -31,6 +34,22 @@ using namespace std;
 using namespace dev;
 using namespace yul;
 
+void CommonSubexpressionEliminator::run(Dialect const& _dialect, Block& _ast)
+{
+	map<YulString, SideEffects> functionSideEffects =
+		SideEffectsPropagator::sideEffects(_dialect, CallGraphGenerator::callGraph(_ast));
+	CommonSubexpressionEliminator cse{_dialect, functionSideEffects};
```
Comment (ekpyron) [inline libyul/optimiser/CommonSubexpressionEliminator.cpp]:
```suggestion
	CommonSubexpressionEliminator cse{_dialect, std::move(functionSideEffects)};
```

File: libyul/optimiser/CommonSubexpressionEliminator.cpp [hunk 4]
```diff
@@ -31,6 +34,22 @@ using namespace std;
...
+	cse(_ast);
+}
+
+CommonSubexpressionEliminator::CommonSubexpressionEliminator(
+	Dialect const& _dialect,
+	map<YulString, SideEffects> const& _functionSideEffects
```
Comment (ekpyron) [inline libyul/optimiser/CommonSubexpressionEliminator.cpp]:
```suggestion
	map<YulString, SideEffects> _functionSideEffects
```

File: libyul/optimiser/CommonSubexpressionEliminator.cpp [hunk 5]
```diff
@@ -31,6 +34,22 @@ using namespace std;
...
+):
+	DataFlowAnalyzer(_dialect, &_functionSideEffects)
```
Comment (ekpyron) [inline libyul/optimiser/CommonSubexpressionEliminator.cpp]:
```suggestion
	DataFlowAnalyzer(_dialect, std::move(_functionSideEffects))
```

File: libyul/optimiser/CommonSubexpressionEliminator.h [hunk 6]
```diff
@@ -37,7 +38,14 @@ struct Dialect;
...
 {
 public:
-	CommonSubexpressionEliminator(Dialect const& _dialect): DataFlowAnalyzer(_dialect) {}
+	/// Runs the CSE pass. @a _ast needs to be the complete AST of the program!
+	static void run(Dialect const& _dialect, Block& _ast);
+
+private:
+	CommonSubexpressionEliminator(
+		Dialect const& _dialect,
+		std::map<YulString, SideEffects> const& _functionSideEffects
```
Comment (ekpyron) [inline libyul/optimiser/CommonSubexpressionEliminator.h]:
```suggestion
		std::map<YulString, SideEffects> _functionSideEffects
```

File: libyul/optimiser/DataFlowAnalyzer.h [hunk 7]
```diff
@@ -67,8 +68,16 @@ struct Dialect;
...
 {
 public:
-	explicit DataFlowAnalyzer(Dialect const& _dialect):
+	/// @param _functionSideEffects
+	///            Side-effects of user-defined functions. Worst-case side-effects are assumed
+	///            if this is not provided or the function is not found.
+	///            The parameter is mostly used to determine movability of expressions.
+	explicit DataFlowAnalyzer(
+		Dialect const& _dialect,
+		std::map<YulString, SideEffects> const* _functionSideEffects = nullptr
```
Comment (ekpyron) [inline libyul/optimiser/DataFlowAnalyzer.h]:
```suggestion
		std::map<YulString, SideEffects> _functionSideEffects = {}
```

File: libyul/optimiser/DataFlowAnalyzer.h [hunk 8]
```diff
@@ -67,8 +68,16 @@ struct Dialect;
...
+	):
 		m_dialect(_dialect),
+		m_functionSideEffects(_functionSideEffects),
```
Comment (ekpyron) [inline libyul/optimiser/DataFlowAnalyzer.h]:
```suggestion
		m_functionSideEffects(std::move(_functionSideEffects)),
```

File: libyul/optimiser/DataFlowAnalyzer.h [hunk 9]
```diff
@@ -124,6 +133,9 @@ class DataFlowAnalyzer: public ASTModifier
 	) const;
 
 	Dialect const& m_dialect;
+	/// Side-effects of user-defined functions. Worst-case side-effects are assumed
+	/// if this is not provided or the function is not found.
+	std::map<YulString, SideEffects> const* m_functionSideEffects = nullptr;
```
Comment (ekpyron) [inline libyul/optimiser/DataFlowAnalyzer.h]:
```suggestion
	std::map<YulString, SideEffects> m_functionSideEffects;
```

File: libyul/optimiser/UnusedPruner.cpp [hunk 10]
```diff
@@ -120,26 +130,31 @@ void UnusedPruner::runUntilStabilised(
...
 	Dialect const& _dialect,
 	Block& _ast,
 	set<YulString> const& _externallyUsedFunctions
 )
 {
+	map<YulString, SideEffects> functionSideEffects =
+		SideEffectsPropagator::sideEffects(_dialect, CallGraphGenerator::callGraph(_ast));
 	bool allowMSizeOptimization = !MSizeFinder::containsMSize(_dialect, _ast);
-	runUntilStabilised(_dialect, _ast, allowMSizeOptimization, _externallyUsedFunctions);
+	runUntilStabilised(_dialect, _ast, allowMSizeOptimization, &functionSideEffects, _externallyUsedFunctions);
```
Comment (ekpyron) [inline libyul/optimiser/UnusedPruner.cpp:157]: Here as well - why not move down by value and default-initifalize as empty map in the constructor?

Comment (chriseth) [reply libyul/optimiser/UnusedPruner.cpp:157]: I guess the thought was that we might re-use the map for multiple stages with re-computing, but it turned out that this is not the case.

Comment (ekpyron) [reply libyul/optimiser/UnusedPruner.cpp:157]: Yeah, we can keep it as it is as well - no optimizer step should ever make side effects of any function *worse*, so it'd always be safe to re-use the map...

Comment (chriseth) [reply libyul/optimiser/UnusedPruner.cpp:157]: I'll keep it like that here, because it will be called in a loop.

Comment (ekpyron) [reply libyul/optimiser/UnusedPruner.cpp:157]: Ah right - I missed the loop, right.

**Review by ekpyron — APPROVED** [review]
I haven't checked the eWasm test update - and we could squash this - looks good to me now in any case!

**Pull request merged**